Add React and React-router

bobbiec opened the pull request:
This PR adds React with react-router.

![image](https://user-images.githubusercontent.com/17092390/52507651-24a4b500-2bc0-11e9-8bdc-c21d0804e016.png)

I basically just deleted everything we had before and ran create-react-app, and then added react router and a little sample code.

bobbiec committed:
Initial commit of create-react-app typescript

bobbiec committed:
Explicitly use react-dev-utils so Travis will run

bobbiec committed:
Explicitly use cross-spawn so maybe Travis will run

bobbiec committed:
Regenerate package-lock.json in hopes that Travis will work this time

bobbiec committed:
Remove package-lock for Travis

bobbiec committed:
Add react router

ykotturi commented on package.json:
```diff
@@ -1,46 +1,37 @@
-    "type": "git",
-    "url": "git+https://github.com/CMU-17-356/dronuts-2019-group-5.git"
+  "version": "0.1.0",
+  "private": true,
```
Why did we change the repo to be private?

bobbiec replied:
This is entirely because I deleted old package.json to run CRA. Good catch!

ykotturi commented on src/tests/App.test.tsx:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
```
Looks like we moved test directory from root to src in this PR - this is probably wise: http://www.wisdomofjim.com/blog/why-i-recommend-unit-tests-in-the-src-folder just want to make sure this was intentionally done...

bobbiec replied:
This was not a conscious decision, just me following different directory structures in different tutorials at different points in time. I think it's fairly arbitrary, though, as long as we stick to one.

ykotturi commented on tslint.json:
```diff
@@ -1,60 +1,10 @@
-        "no-null-keyword": true,
-        "prefer-const": true,
-        "jsdoc-format": true
+    "extends": [],
```
I don't understand what happened to linting here... it looks like we removed the core functionality - maybe this is added in later commit? Or will become clearer in later commit? WIll check back...

bobbiec replied:
I think this is actually an artifact of me messing around with manually putting all the parts together (before I just went to using create-react-app). CRA itself doesn't support typescript ([link](https://github.com/facebook/create-react-app/issues/5641)) right now, so I'm going to remove this. Linting is provided by ESLint instead.

ykotturi commented on package.json:
```diff
@@ -30,6 +30,7 @@
   "devDependencies": {
     "@types/enzyme": "^3.1.17",
     "@types/enzyme-adapter-react-16": "^1.0.3",
+    "cross-spawn": "^6.0.5",
```
$ npm install cross-spawn

ykotturi commented on .gitignore:
```diff
@@ -1,3 +1,7 @@
+# Because otherwise Travis tests will fail
+# Todo: look into this
+package-lock.json
```
I couldn't comment on your prior commit as it as deleting a file (package-lock.json) - but It's unclear to me why deleting this file was necessary? Rather than regenerating it.. trying to understand your thinking -- would love to hear your thoughts!

bobbiec replied:
When package-lock.json is present, Travis runs `npm ci` rather than `npm install` ([link](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#npm-ci-support)). This causes tests to fail because dependencies are missing. 

I suspect this is because devDependencies are not installed, but I don't know how to fix that. I also don't know how to disable the use of `npm ci` except by omitting `package-lock.json`.

ykotturi commented on src/App.tsx:
```diff
@@ -1,28 +1,10 @@
-import React, { Component } from 'react';
-import logo from './logo.svg';
-import './App.css';
+import * as React from 'react';
```
I'm still going through React tutorial - I'll return after I've finished tutorial to review - unless others beat me to it..

ykotturi commented:
I just pulled this PR branch locally and got it up n' running w/ docker. I'll finish the React tutorial and then return to finish the review (specifically, on the last commit of this PR), which is all React additions.

bobbiec committed:
Address Yasmine's comments

- Remove redundant tslint.json and update README about linting
- Restore package.json details from old version

ykotturi commented:
Ok - I've reviewed your React code as well as tested locally. All looks good - although I have some questions about routing in general, which hopefully we can spend ~5 minutes on when you arrive for group meeting - if that's OK with you @bobbiec. 

Merging this branch.

ykotturi commented:
Oh wait, can you update your branch w/ latest master locally, and then push? It's a few commits behind. @bobbiec  And then I'll merge. Thanks.

bobbiec committed:
Merge branch 'master' of github.com:CMU-17-356/dronuts-2019-group-5 into create-react-app

ykotturi commented:
When we rerun each of these Travis builds independently, they work. Going to merge to master!
